hkmc2/shared/src/test/mlscript-compile/apps/parsing-web-demo/main.mls

```diff
@@ -0,0 +1,223 @@
+    let white_part = spaces (largest_disk_size + 1 - n) in
+    white_part ^ (disk n) ^ white_part;;
+let peg_base largest_disk_size =
+    let half = make_string largest_disk_size "_" in
+    " " ^ half ^ "|" ^ half ^ " ";;
+let rec peg largest_disk_size = function
+  | (0, []) -> []
+  | (0, head::rest) ->
+      disk_number head largest_disk_size ::
+      peg largest_disk_size (0, rest)
+  | (offset, lst) ->
+      disk_number 0 largest_disk_size ::
+      peg largest_disk_size (offset-1, lst);;
+let rec join_lines l1 l2 l3 =
+  match (l1, l2, l3) with
+  | ([], [], []) -> []
+  | (t1::r1, t2::r2, t3::r3) -> (t1 ^ t2 ^ t3) :: join_lines r1 r2 r3
+  | _ -> failwith "join_lines";;
+"""
+
+fun showExample2() =
+  set editor.value = """#newKeyword ("hello", Some 3, Some 3)
+#newKeyword ("goodbye", None, None)
+  
+#newCategory("greeting")
+  
+#extendCategory("greeting", [ keyword("hello"), "term", "greeting" ], foo)
+#extendCategory("greeting", [ keyword("goodbye") ], bar)
+  
+#extendCategory("decl", [ "greeting" ], baz)
+  
+  
+hello "Rob" hello "Bob" goodbye
+  
+#diagram ""
+"""
+
+showExample1()
+
+editor.addEventListener of "keydown", event =>
+  if event.key is "Tab" do
+    event.preventDefault()
+    let start = editor.selectionStart
+    let end = editor.selectionEnd
+    set editor.value = editor.value.substring(0, start) + "  " + editor.value.substring(end)
+    set editor.selectionEnd = start + 2
+    set editor.selectionStart = editor.selectionEnd
+
+selector.addEventListener of "change", event => if selector.value is
+  "example-1" do showExample1()
+  "example-2" do showExample2()
+
+parseButton.addEventListener of "click", event =>
+  let tokens = Lexer.lex(editor.value, noWhitespace: true)
+  set outputPanel.innerHTML = ""
+  Runtime.try_catch of
+    () =>
+      let trees = Parser.parse(tokens)
+      trees Iter.fromStack() Iter.each of tree => if
+        Extension.isDiagramDirective(tree) then displayRules()
+        else
+          let collapsibleTree = document.createElement("collapsible-tree")
+          set collapsibleTree.textContent = TreeHelpers.showAsTree(tree)
+          outputPanel.appendChild of collapsibleTree
+    error =>
+      let errorDisplay = document.createElement("error-display")
+      errorDisplay.setError of error
+      outputPanel.appendChild of errorDisplay
+
+let indentRegex = new RegExp("""^(\s*)""")
+
+fun parseIndentedText(text) =
+  let root = (text: "", children: [])
+  let stack = [(node: root, indent: -1)]
+  text.split("\n")
+    Iter.filtering(line => line.trim().length > 0)
+    Iter.each of line =>
+      let indent = line.match(indentRegex).[1].length
+      let text = line.substring(indent)
+      while indent <= stack.[stack.length - 1].indent do
+        stack.pop()
+      let newNode = (text: text, children: [])
+      stack.[stack.length - 1].node.children.push of newNode
+      stack.push of node: newNode, indent: indent
+  root.children
+
+let errorDisplayStyle = """
```

Comment:
Any reason this is not defined in the HTML?

Reply:
Yes, for sure it's better than anything complicated. Please document what you wrote above in a code comment, though.